tls/test_tls_tickets.py

```diff
@@ -0,0 +1,319 @@
+        ticket = getattr(hs.sock.tls_ctx, 'ticket', None)
+        self.assertIsNotNone(ticket,
+                             'Ticket value is empty, no NewSessionTicket message '
+                             'was found')
+        self.assertNotEqual(ticket.ticket, t_random_data,
+                            'No new ticket was provided by Tempesta')
+
+        # Now insert a modification into valid ticket data, ticket must be
+        # rejected.
+        t_data = ticket.ticket
+        index = len(t_data) / 2
+        repl = 'a' if t_data[index] != 'a' else 'b'
+        t_data = t_data[:index] + repl + t_data[index + 1:]
+        hs = TlsHandshake()
+        hs.set_ticket_data(t_data)
+        hs.session_id = '\x39' * 32
+        res = hs.do_12()
+        self.assertTrue(res, "Wrong handshake result: %s" % res)
+        ticket = getattr(hs.sock.tls_ctx, 'ticket', None)
+        self.assertIsNotNone(ticket,
+                             'Ticket value is empty, no NewSessionTicket message '
+                             'was found')
+        self.assertNotEqual(ticket.ticket, t_random_data,
+                            'No new ticket was provided by Tempesta')
+
+
+class TlsVhostConfusion(tester.TempestaTest):
+
+    backends = [
+        {
+            'id' : '0',
+            'type' : 'deproxy',
+            'port' : '8000',
+            'response' : 'static',
+            'response_content' :
+                'HTTP/1.1 200 OK\r\n'
+                'Content-Length: 0\r\n'
+                'X-Vhost: tempesta-tech\r\n'
+                'Connection: keep-alive\r\n\r\n'
+        },
+        {
+            'id' : '1',
+            'type' : 'deproxy',
+            'port' : '8001',
+            'response' : 'static',
+            'response_content' :
+                'HTTP/1.1 200 OK\r\n'
+                'Content-Length: 0\r\n'
+                'X-Vhost: tempesta\r\n'
+                'Connection: keep-alive\r\n\r\n'
+        }
+    ]
+
+    tempesta = {
+        'custom_cert': True,
+        'config' : """
+            cache 0;
+            listen 443 proto=https;
+
+            srv_group srv_grp1 {
+                server ${server_ip}:8000;
+            }
+            srv_group srv_grp2 {
+                server ${server_ip}:8001;
+            }
+
+            vhost tempesta-tech.com {
+                proxy_pass srv_grp1;
+                tls_certificate ${tempesta_workdir}/tempesta-tech.com.crt;
+                tls_certificate_key ${tempesta_workdir}/tempesta-tech.com.key;
+            }
+            vhost tempesta.com {
+                proxy_pass srv_grp1;
+                tls_certificate ${tempesta_workdir}/tempesta.com.crt;
+                tls_certificate_key ${tempesta_workdir}/tempesta.com.key;
+            }
+            http_chain {
+                host == "tempesta-tech.com" -> tempesta-tech.com;
+                host == "tempesta.com" -> tempesta.com;
+                -> block;
+            }
```

Comment:
It makes sense to test the behavior with a single global certificate and the two vhosts - in this configuration we should be good to establish a TLS connection.

Another test case which we also discussed is a vhost and a default vhost.

Please don't forget to describe the behavior (with the reference to the paper) in wiki.

Reply:
Extra scenarios was added. Unfortunately, vhost confusion happens in new `TlsVhostConfusionDfltCerts` and `TlsVhostConfusionDfltCertsWithUnknown` scenarios - working on them.